File: JavaScript/src/Internals/Functions/MathSum/Function_TDIST.js
import { Function_3 } from '../Function_3.js';
import { Operand } from '../../../Operand.js';
import { ExcelFunctions } from '../../../MathNet/ExcelFunctions.js';
import { StringCache } from '../../../Internals/StringCache.js';

class Function_TDIST extends Function_3 {
    constructor(z) {
    super(z);
  }

    Evaluate(engine, tempParameter) {
        let args1 = this.a.Evaluate(engine, tempParameter);
            args1 = args1.ToNumber(StringCache.Function_parameter_error, "TDist", 1);
            if (args1.IsError) return args1;
            
        let args2 = this.b.Evaluate(engine, tempParameter);
            args2 = args2.ToNumber(StringCache.Function_parameter_error, "TDist", 2);
            if (args2.IsError) return args2;
            
        let args3 = this.c.Evaluate(engine, tempParameter);
            args3 = args3.ToNumber(StringCache.Function_parameter_error, "TDist", 3);
            if (args3.IsError) return args3;
            
        let x = args1.NumberValue;
        let degreesFreedom = args2.IntValue;
        let tails = args3.IntValue;
        if (degreesFreedom <= 0.0 || tails < 1 || tails > 2) {
            return Operand.Error(StringCache.Function_parameter_error, "TDist");
        }
        return Operand.Create(ExcelFunctions.TDist(x, degreesFreedom, tails));
    }
}

export { Function_TDIST };

File: java/src/main/java/toolgood/algorithm/internals/functions/mathsum/Function_TDIST.java
package toolgood.algorithm.internals.functions.mathsum;

import java.util.function.Function;

import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.Function_3;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.mathNet.ExcelFunctions;


public class Function_TDIST extends Function_3 {
    public Function_TDIST(FunctionBase func1, FunctionBase func2, FunctionBase func3) {
        super(func1, func2, func3);
    }

    @Override
    public Operand Evaluate(Object work, Function<Object, String, Operand> tempParameter) {
        Operand args1 = func1.Evaluate(work, tempParameter);
        if (args1.getOperandType() != OperandType.Number) {
            args1 = args1.ToNumber
("Function '{0}' parameter {1} is error!", "TDist", 1);
            if (args1.IsError()) {
                return args1;
            }
        }
        Operand args2 = func2.Evaluate(work, tempParameter);
        if (args2.getOperandType() != OperandType.Number) {
            args2 = args2.ToNumber
("Function '{0}' parameter {1} is error!", "TDist", 2);
            if (args2.IsError()) {
                return args2;
            }
        }
        Operand args3 = func3.Evaluate(work, tempParameter);
        if (args3.getOperandType() != OperandType.Number) {
            args3 = args3.ToNumber
("Function '{0}' parameter {1} is error!", "TDist", 3);
            if (args3.IsError()) {
                return args3;
            }
        }

        double x = args1.DoubleValue
();
        int degreesFreedom = args2.IntValue();
        int tails = args3.IntValue();

        if (degreesFreedom <= 0 || tails < 1 || tails > 2) {
            return Operand.Error("Function '{0}' parameter is error!", "TDist");
        }

        return Operand.Create(ExcelFunctions.TDist(x, degreesFreedom, tails));
    }

    @Override
    public void toString(StringBuilder stringBuilder, boolean addBrackets) {
        AddFunction(stringBuilder, "TDist");
    }
}
